control_panel_api/tests/test_views.py

```diff
@@ -20,6 +20,26 @@
 )
 
 
+class AssertionsMixin(object):
+    def assert_s3bucket(self, s3bucket):
+        self.assertIn('id', s3bucket)
+        self.assertIn('url', s3bucket)
+        self.assertIn('name', s3bucket)
+        self.assertIn('arn', s3bucket)
+        self.assertIn('created_by', s3bucket)
+        self.assertEqual(5, len(s3bucket))
```

Comment:
With this assertion, the previous ones add up to something like `self.assertEqual(s3bucket.field_names, expected_field_names)` - can it be reduced to a single assertion?

Reply:
That doesn't read as clear to me.

Reply:
Just minor advantages - it would be more concise (fewer lines to read and fewer places for bugs to hide) and it would reveal the intent (assert that the object has only the fields we expect)

Reply:
What I mean about revealing the intent is, this method _says_ it asserts that the passed object is a `s3bucket`, but what it _does_ is assert that the object has only the five fields we specify.

The intention of the tests that call it appears to be to verify that the JSON objects in the response represent an `S3Bucket` model - maybe the assertion should be that deserializing the JSON object to an instance of S3Bucket succeeds?

Reply:
Well my intention is just to assert the correct fields I would expect back when serializing the app or bucket.  So something like `assert_s3bucket_representation` etc etc?

I do not care about the data in these cases, as that would make the test brittle.  I just want to know that the correct fields have been returned due to serializer usage.

Reply:
Regarding a single assertion it could look like this:

```
expected_fields = {
    'id',
    'url',
    'name',
    'arn',
    'created_by'
}
self.assertEqual(expected_fields, set(s3bucket))
```

So the question is which is more readable?  @xoen

Reply:
I wouldn't mind, as far as I can tell they both read well enough to me...(but I see Andy's point about the data).

Reply:
It's more the list of hard-coded field names that bothers me - you will have to edit both the model and the tests when you make changes to the fields.

This method would better match your stated intention if it checked that the JSON is a serialized `S3Bucket`, maybe by deserializing it, or by comparing the field names to a freshly serialized one, or something else.

Reply:
Mmmm, I don't know if it's a good idea to get this list implicitly from the model. I'd prefer to be explicit at the small cost of having to add/remove an item from the list.

Plus, we (for whatever reason) may want to hide some of the fields in the model.

Just my 2p.

Reply:
Had a rethink of the serializer field testing - totally forgot about the `Serializer.Meta.fields` var.  I refactored all the tests to use a nice, explicit field assertion pattern.  I also refactored a couple tests that can be included in existing test methods.  It also eliminated the `AssertionsMixin`.  I've pushed up a commit which I'm happy with.

Reply:
I've played with using the Serializer.Meta.fields var, which works, but Aldo is right, the fields should be specified and tested because we should use the test as a contract for the representation - if someone changes the serializer fields accidentally we need to see the test blow up.  If you reduce it to dynamic one liners the tests will always pass, I could just delete a load of important fields and we wouldn't know.